### PR TITLE
expose additional container and compose functionality, allow setting the mine function

### PR DESCRIPTION
**Pull request by fraajad**
Hi, I found this formula really useful so I added some functionality.

This pull request...
1. maintains backward compatibility
2. container.sls now exposes all dockerng.running parameters
3. container.sls makes volume directory creation optional so single files can be mounted (ie: /var/run/docker.sock)
4. docker-compose.yml
  a. is rendered in pure YAML
  b. adds support for volumes key
  c. allows idiomatic "services" (and "networks") but supports "service" key
  d. supports empty key volume definitions (ie data:)
5. swarm.sls allows setting a different mine.get function.  grains.items can expose sensitive data so grains.item may be used instead.

Cheers,
Doug

```
###
# Use case example
###


docker:
  host:
    enabled: true
    pkgs:
      - docker-engine=17.05.0*
    options:
      bip: 192.168.129.1/24
      log-driver: json-file
      log-opts:
        max-size: 500m
      dns: 
        - 192.168.129.1
      dns-search: 
        - service.consul
  client:
    enabled: true
    compose:
      source:
        engine: docker
        image: docker/compose:1.14.0

    container:
      registrator:
        image: gliderlabs/registrator:v7
        command: -ip {{ salt.grains.get('ip_interfaces:eth0:0') }} -deregister="on-success" consul://localhost:8500
        network_mode: host
        volumes:
          - /var/run/docker.sock:/tmp/docker.sock
        makedirs: False
        restart_policy: on-failure

      logspout:
        image: vincit/logspout-gelf
        command: gelf://graylog.service.consul:12201
        hostname: {{ grains['id'] }}
        environment:
          BACKLOG: '"false"'
        ports:
          - 8000:80
        volumes:
          - /var/run/docker.sock:/var/run/docker.sock
        makedirs: False
        restart_policy: on-failure

    compose:
      graylog:
        # Run up action, any positional argument to docker-compose CLI
        # If not defined, only docker-compose.yml is generated
        status: up
        # Run image pull every time state is run triggering container
        # restart in case it's changed
        pull: false
        volumes:
          mongo-db:
          mongo-configdb:
          elasticsearch-data:
          graylog-data:
        service:
          mongo:
            image: mongo:3
            restart: unless-stopped
            network_mode: bridge
            volumes:
              - mongo-db:/data/db
              - mongo-configdb:/data/configdb
          elasticsearch:
            image: elasticsearch:2
            command: elasticsearch -Des.cluster.name='graylog'
            network_mode: bridge
            restart: unless-stopped
            volumes:
              - elasticsearch-data:/usr/share/elasticsearch/data
          server:
            image: graylog2/server:2.1.1-1
            network_mode: bridge
            dns: 
              - 192.168.129.1
            restart: unless-stopped
            volumes:
              - graylog-data:/usr/share/graylog/data
            environment:
              GRAYLOG_PASSWORD_SECRET: somepasswordpepper
              GRAYLOG_ROOT_PASSWORD_SHA2: 8c6976e5b5410415bde908bd4dee15dfb167a9c873fc4bb8a81f6f2ab448a918
              SERVICE_9000_NAME: graylog
              SERVICE_12201_NAME: gelf-udp
            links:
              - mongo
              - elasticsearch
            ports:
              - 9000:9000
              - 12201:12201/udp
              
```

### CODE REVIEW

**Comment by cloudlab-autobuild** [comment]
Can one of the admins verify this patch?

File: docker/map.jinja [hunk 1]
```diff
@@ -22,7 +22,8 @@
     'default': {
```
Comment (fpytloun) [inline docker/map.jinja:22]: Don't need to change map.jinja mode (add execute bit), leave it 644

**Review by fpytloun — CHANGES_REQUESTED** [review]
Thank you for your contribution, I like the change :-)

Ad. singular vs. plural in pillar, we use following convention:
- singular for dictionary (eg. service) because it makes more sense to refer service.mysvc (so you can naturally say "service mysvc")
- plural for lists (eg. services) so you can naturally say "first item of list of services"

So I would prefer keeping this convention.

**Comment by fraajad** [comment]
Right on!  No problem, the naming convention makes sense.  And the changes have been made.

File: docker/map.jinja (hunk 1 above)
Comment (fraajad) [reply docker/map.jinja:22]: Fixed, with a couple other small corrections

**Pull request merged**